test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -51,258 +53,475 @@ var (
+				// this more intuitive?
+				defer nodePreparer.CleanupNodes()
+			}
+			numNodes += realOp.CreateNodes
+
+		case *createPodsOp:
+			var namespace string
+			if realOp.Namespace != nil {
+				namespace = *realOp.Namespace
+			} else {
+				namespace = fmt.Sprintf("namespace-%d", opIndex)
+			}
+			var stopCh chan struct{}
+			var collectors []testDataCollector
+			if realOp.CollectMetrics {
+				b.StartTimer()
+				stopCh = make(chan struct{})
+				collectors = getTestDataCollectors(podInformer, fmt.Sprintf("%s/%s", b.Name(), namespace), namespace, tc.MetricsCollectorConfig)
+				for _, collector := range collectors {
+					go collector.run(stopCh)
+				}
+			}
+			if err := createPods(namespace, realOp, clientset); err != nil {
+				b.Fatal(err)
+			}
+			finish := func() {
+				if err := barrierOne(podInformer, b.Name(), namespace, realOp.CreatePods); err != nil {
+					b.Fatal(err)
+				}
+				if realOp.CollectMetrics {
+					close(stopCh)
+					mu.Lock()
+					for _, collector := range collectors {
+						dataItems = append(dataItems, collector.collect()...)
+					}
+					mu.Unlock()
+				}
+			}
+			if realOp.SkipWaitToCompletion {
+				// Only record those namespaces that may potentially require barriers
+				// in the future.
+				if _, ok := numPodsScheduledPerNamespace[namespace]; ok {
+					numPodsScheduledPerNamespace[namespace] += realOp.CreatePods
+				} else {
+					numPodsScheduledPerNamespace[namespace] = realOp.CreatePods
+				}
+				go finish()
+				// We can't stop the timer after the pods actually get scheduled
+				// because a different timer may have been started and stopped by then.
+				// AFAIK, there is no way to add the time elapsed to the benchmark's
+				// timer concurrently without using StartTimer and StopTimer.
+				b.StopTimer()
+			} else {
+				allPodsAsync = false
+				finish()
+				b.StopTimer()
+			}
 
-	// Start test data collectors.
-	stopCh := make(chan struct{})
-	collectors := getTestDataCollectors(test, podInformer, b)
-	for _, collector := range collectors {
-		go collector.run(stopCh)
+		case *barrierOp:
+			for _, barrier := range realOp.Barrier {
+				if _, ok := numPodsScheduledPerNamespace[barrier]; !ok {
+					b.Fatalf("unknown namespace %s", barrier)
+				}
+			}
+			if err := barrier(podInformer, b.Name(), realOp.Barrier, numPodsScheduledPerNamespace); err != nil {
+				b.Fatal(err)
+			}
+			// At the end of the barrier, we can be sure that there are no pods
+			// pending scheduling in the namespaces that we just blocked on.
+			if len(realOp.Barrier) == 0 {
+				numPodsScheduledPerNamespace = make(map[string]int)
+			} else {
+				for _, namespace := range realOp.Barrier {
+					delete(numPodsScheduledPerNamespace, namespace)
+				}
+			}
+		default:
+			b.Fatalf("invalid op %v", realOp)
+		}
 	}
 
-	// Schedule the main workload
-	if err := createPods(testNamespace, test.PodsToSchedule, clientset); err != nil {
-		b.Fatal(err)
+	if allPodsAsync {
```

Comment:
ditto, I think this is not necessary, stop/start from beginning to end of `runWorkload`.

Reply:
Done.